### ISSUE TITLE
UI Opacity

### ISSUE THREAD

**Issue by fianathedevgirl** [bevyengine/bevy#6956]
## What problem does this solve or what need does it fill?

This allows fade-in fade-out transitions for UI bundles, and setting the overall transparency of UI bundles, which I think is useful in many games to unclutter the window when it's not really important (but may be needed) or as a configuration in games for players who wish to unclutter their view and see more what's going on.

## What solution would you like?

To be discussed.

## What alternative(s) have you considered?

For my own game, I am considering trying to animate with [`Transform`](https://docs.rs/bevy/latest/bevy/prelude/struct.Transform.html). It allows a more cartoony transition, but for others this might not been what they prefer for their game theme.

It's possible to change the opacity of [`TextBundle`](https://docs.rs/bevy/latest/bevy/prelude/struct.TextBundle.html)s and background colors but that's a major hack and can't be used with [`ImageBundle`](https://docs.rs/bevy/latest/bevy/prelude/struct.ImageBundle.html)

## Additional context

https://developer.mozilla.org/en-US/docs/Web/CSS/opacity

**Comment by viridia** [comment]
The way I would like to see this done is with buffered rendering. So for example, if you have an animated dialog that needs to transition (fade-in, scale, etc.) it would be rendered to an offscreen render target, and then the render target would be rendered to the main view. This allows you to control the opacity of the entire dialog, instead of making the children individually translucent (which would look bad).

I propose that there be some Component that you attach to a UI node that says "render all of the children of this node onto this offscreen rendertarget", at which point the node itself simply sets that same render target as its background image. You can then either introduce a new Opacity style property, or use the existing behavior where the alpha component of the background color is multiplied with the background image.

**Comment by artswithastriya** [comment]
> This allows you to control the opacity of the entire dialog, instead of making the children individually translucent (which would look bad).

Yeah. I'm not exactly sure if this is accurate but generally in raster programs, opacity in folders works but occasionally normal mode makes it better?

**Comment by pjgf94** [comment]
The way i would like this implemented is (like Godot does). By using a Modulate, SelfModulate components:
```
#[derive(Component, Default)]
#[default(0: Color::WHITE)] // crate better_default
pub struct Modulate(Color);

#[derive(Component, Default)]
#[default(0: Color::WHITE)]
pub struct SelfModulate(Color);

#[derive(Component)]
pub struct CalculatedModulate(vec4);
```
Where CalculatedModulate gets updated the same way as Transform, by multiplying with the modulate of the parent entity if any, and then the calculated modulate gets multiplied with the self modulate(after converting to the same color space), and that color gets multiplied in the shader with the output color of the pixel.

This way can be reused in 2d, 3d and ui, and can also be used for simple effects like for example:
- tinting the character green when poisoned, red when bleeding
- giving semitransparency to a 3d character composed of various hierarchical entities, for a ghost effect.
